src/custom/int.rs

```diff
@@ -25,18 +27,78 @@ use anyhow::{anyhow, Result};
 
 struct IntOpEmitter<'c, 'd, H: HugrView>(&'d mut EmitFuncContext<'c, H>);
 
+// TODO this is probably useful enough to offer as a general utility
+fn emit_custom_binary_op<'c, H: HugrView>(
+    context: &mut EmitFuncContext<'c, H>,
+    args: EmitOpArgs<'c, CustomOp, H>,
+    go: impl FnOnce(
+        &Builder<'c>,
+        BasicValueEnum<'c>,
+        BasicValueEnum<'c>,
+    ) -> Result<Vec<BasicValueEnum<'c>>>,
+) -> Result<()> {
+    let [lhs, rhs] = TryInto::<[_; 2]>::try_into(args.inputs).map_err(|v| {
+        anyhow!(
+            "emit_custom_2_to_1_op: expected exactly 2 inputs, got {}",
+            v.len()
+        )
+    })?;
+    if lhs.get_type() != rhs.get_type() {
+        return Err(anyhow!(
+            "emit_custom_2_to_1_op: expected inputs of the same type, got {} and {}",
+            lhs.get_type(),
+            rhs.get_type()
+        ));
+    }
+    let res = go(context.builder(), lhs, rhs)?;
+    if res.len() != args.outputs.len()
+        || zip_eq(res.iter(), args.outputs.get_types()).any(|(a, b)| a.get_type() != b)
+    {
+        return Err(anyhow!(
+            "emit_custom_2_to_1_op: expected outputs of types {:?}, got {:?}",
+            args.outputs.get_types().collect_vec(),
+            res.iter().map(BasicValueEnum::get_type).collect_vec()
+        ));
+    }
+    args.outputs.finish(context.builder(), res)
+}
+
+fn emit_icmp<'c, H: HugrView>(
+    context: &mut EmitFuncContext<'c, H>,
+    args: EmitOpArgs<'c, CustomOp, H>,
+    pred: inkwell::IntPredicate,
+) -> Result<()> {
+    let true_val = emit_value(context, &Value::true_val())?;
+    let false_val = emit_value(context, &Value::false_val())?;
+
+    emit_custom_binary_op(context, args, |builder, lhs, rhs| {
+        // get result as an i1
+        let r = builder.build_int_compare(pred, lhs.into_int_value(), rhs.into_int_value(), "")?;
+        // convert to whatever BOOL_T is
+        Ok(vec![builder.build_select(r, true_val, false_val, "")?])
```

Comment:
Aside: Have you considered lowering unit sums directly to ints? Then you wouldn't need this select and could directly switch for branching?

Reply:
> What we do now is in effect lowering to i32, on the assumption that all the nested struct nonsense get's deleted by an optimiser. [...] My expectation is that the output of optimisation will be the same as if we represented BOOL_T as i1 directly

Ah I see. Agreed, if LLVM can deal with it then your simple solution is better 👍